Bayesian_optimization/acq.py

```diff
@@ -0,0 +1,311 @@
+import inspect
+import math
+from numbers import Real
+import numpy as np
+from scipy.stats import norm
+
+def acq_optimize(self, niteration, lr):
+    optimizer = torch.optim.Adam(self.parameters(), lr=lr)
+    # optimizer.zero_grad()
+    for i in range(niteration):
+        optimizer.zero_grad()
+        loss = self.negative_acq()
+        loss.backward()
+        optimizer.step()
+        print('iter'+str(i)+'/'+str(niteration), 'loss_negative_acq:', loss.item(), end='\r')
+
+
+def find_next_batch(acq, bounds, batch_size=1, n_samples=1000, f_best=0):
+    """
+    Find the next batch of points to sample by selecting the ones with the highest UCB from a large set of random samples.
+
+    Args:
+        bounds (np.ndarray): The bounds for each dimension of the input space.
+        batch_size (int): The number of points in the batch.
+        n_samples (int): The number of random points to sample for finding the maximum UCB.
+
+    Returns:
+        torch.Tensor: The next batch of points to sample.
+    """
+
+    # 获取forward函数的参数信息
+    signature = inspect.signature(acq.forward)
+    parameters = signature.parameters
+
+    # 打印参数数量
+    num_params = len(parameters)
+    print(f"Number of parameters in forward function: {num_params}")
+
+    X_selected = []
+    for _ in range(batch_size):
+        # Generate a large number of random points
+        X_random = torch.FloatTensor(n_samples, bounds.shape[0]).uniform_(bounds[0, 0], bounds[0, 1])
+
+        if num_params == 1:
+        # Compute UCB for all random points
+            UCB_values = acq.forward(X_random)
+        if num_params == 2:
+            UCB_values = acq.forward(X_random, f_best)
+        # Select the point with the highest UCB value
+        idx_max = torch.argmax(UCB_values)
+        X_selected.append(X_random[idx_max])
+    return torch.stack(X_selected)
+
+
+class UCB:
+    def __init__(self, mean_func, variance_func, kappa=2.0):
+        """
+        Initialize the Batch Upper Confidence Bound (UCB) acquisition function.
+
+        Args:
+            mean_func (callable): Function to compute the mean of the GP at given points (PyTorch tensor).
+            variance_func (callable): Function to compute the variance of the GP at given points (PyTorch tensor).
+            kappa (float): Controls the balance between exploration and exploitation.
+        """
+        self.mean_func = mean_func
+        self.variance_func = variance_func
+        self.kappa = kappa
+    # forward
+    def forward(self, X):
+        """
+        Compute the UCB values for the given inputs.
+
+        Args:
+            X (torch.Tensor): The input points where UCB is to be evaluated.
+
+        Returns:
+            torch.Tensor: The UCB values for the input points.
+        """
+        mean = self.mean_func(X)
+        variance = self.variance_func(X)
+        return mean + self.kappa * torch.sqrt(variance)
+
+    def find_next_batch(self, bounds, batch_size=1, n_samples=1000):
```

Comment:
TODO: find_next_batch is a common function for all acquisition functions, so it should be moved to a separate file